internal/operands/common-templates/reconcile.go

```diff
@@ -0,0 +1,151 @@
+package common_templates
+
+import (
+	"fmt"
+	templatev1 "github.com/openshift/api/template/v1"
+	core "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"kubevirt.io/ssp-operator/internal/common"
+	"kubevirt.io/ssp-operator/internal/operands"
+	"path/filepath"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sync"
+)
+
+var (
+	loadTemplatesOnce sync.Once
+	templatesBundle   []templatev1.Template
+)
+
+type commonTemplates struct{}
+
+func GetOperand() operands.Operand {
+	return &commonTemplates{}
+}
+
+func (c *commonTemplates) AddWatchTypesToScheme(s *runtime.Scheme) error {
+	return templatev1.Install(s)
+}
+
+func (c *commonTemplates) WatchClusterTypes() []runtime.Object {
+	return []runtime.Object{
+		&rbac.ClusterRole{},
+		&rbac.Role{},
+		&rbac.RoleBinding{},
+		&core.Namespace{},
+		&templatev1.Template{},
+	}
+}
+
+func (c *commonTemplates) WatchTypes() []runtime.Object {
+	return nil
+}
+
+func (c *commonTemplates) Reconcile(request *common.Request) error {
+	for _, f := range []func(*common.Request) error{
+		reconcileGoldenImagesNS,
+		reconcileViewRole,
+		reconcileViewRoleBinding,
+		reconcileEditRole,
+		reconcileTemplates,
+	} {
+		if err := f(request); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *commonTemplates) Cleanup(request *common.Request) error {
+	objects := []controllerutil.Object{
+		newGoldenImagesNS(GoldenImagesNSname),
+		newViewRole(GoldenImagesNSname),
+		newViewRoleBinding(GoldenImagesNSname),
+		newEditRole(),
+	}
+	namespace := request.Instance.Spec.CommonTemplates.Namespace
+	for _, template := range templatesBundle {
```

Comment:
When you use 
```
for _, template := range remplatesBundle {
    ...
}
```
A single variable `template` is defined in the scope of the loop (not a new one for each iteration), which means that `&template` returns the same address for all iterations. This eventually causes you to append the same address X times to `objects`, which means you are trying to delete the same template X times (the last template that was assigned to `template`.

To solve this, you can use:
```
for index := range templateBundle {
    ...
    objects = append(objects, &templatesBundle[index])
}
```

Reply:
This change seems to have fixed the cleanup!